Turns off QNode caching by default and makes variable values available during construction

josh146 opened the pull request:
**Context:**

Currently, the way the QNode works uses a method of 'JIT' circuit compilation:

* On the first evaluation call, the `QNode.construct` method is called. This acts to:

  - inspect the function signature and the passed args/kwargs for array size/shape,
  - wraps them in the `Variable` class,
  - runs the custom quantum function,
  - extracts the queued operations and queued expectations,
  - builds up a dictionary of which operations depend on which arguments.

From this point on, the user's function is never called again --- the QNode has cached all the information about the arguments, operations, and expectations to directly query the device.

However, there is a potential (advanced) use case where the user might want to do classical processing on **keyword arguments** within the QNode. This is generally a **safe** procedure, as QNodes, by default, treat keyword arguments as non-differentiable data.

**Description of the Change:**

To allow the above code pattern, the following has been modified:

* The QNode now by default re-evaluates the function/re-constructs the circuit on every execution. The `QNode` class and `qnode` decorator have a new keyword argument `cache`. By default this is set to `False`, the user may set this to `True` if they want to return to the old behaviour.

* The QNode now makes variable values available during construction when the function is called. Previously, it only did this during evaluation, and just used dummy `Variable` classes during construction.

**Benefits:**

This PR makes the following code pattern legal:

```python
@qml.qnode(dev)
def circuit(x, c=None):
    qml.RX(x, wires=0)

    for i in range(c):
        qml.RX(x, wires=i)

    return qml.expval.PauliZ(0)

circuit(0.5, c=2)
```

i.e., you can perform classical processing of keyword arguments inside the QNode --- data that is assumed to be non-differentiable, and programmatically change the circuit structure.

**Possible Drawbacks:** 

* There could be a small speed penalty in having to reconstruct the QNode with every evaluation. However, since the bottleneck would in almost all cases be the quantum simulation/optimization, this should be negligible.

**Related GitHub Issues:** https://discuss.pennylane.ai/t/passing-non-differentiable-arguments-to-qnode/135/14

josh146 commented:
Actually, the more I think about this, I think this is a Bad Idea™, as it may not be obvious to the user that their classical processing **will only run on construction**, and cannot be updated. On that thought, I am closing this PR.

josh146 commented:
I've re-opened this PR, for two reasons:

1. This is an often attempted pattern, and we've had numerous additional tickets involving it. Users don't expect their custom function to only run on first execution, leading to unexpected results.

2. I've resolved my previous comment re: classical processing only running on construction, by introducing the function `qml.eager_mode()`. By running this function, an experimental 'eager mode' is activated, where QNodes now re-construct circuit by re-evaluating the circuit function on every execution. This is specified via the `QNode.eager` class property.

Ideally, turning on this new eager mode should pass all the existing tests. However, this was not the case, enabling me to fix some unexpected bugs in the test suite. After making these changes, the tests pass under both eager and the default non-eager mode.

josh146 commented:
@co9olguy, @smite, @quantshah - the name `eager` used here doesn't really make sense, and it should be renamed before merging. Any suggestion for an alternative name?

Needs to express that turning on this experimental option changes PennyLane from only evaluating the users quantum function once -> evaluating it on every QNode call.

smite commented on pennylane/qnode.py:
```diff
@@ -198,11 +198,13 @@ class QNode:
     """
     # pylint: disable=too-many-instance-attributes
     _current_context = None  #: QNode: for building Operation sequences by executing quantum circuit functions
+    eager = False
```
```suggestion
    eager = False  #: bool: if True, QNode.func will be called to reconstruct the quantum circuit on every execution of the QNode
```

smite commented:
`qml.mutable_qnodes`?
`qml.mutable_circuits`?

I'm not sure mutable is the proper word here, since it's not mutability in the exact Python sense. The idea is that the if the setting is True, the circuit structure of the QNodes can change on every execution.

smite commented on pennylane/qnode.py:
```diff
@@ -461,9 +467,26 @@ def evaluate(self, args, **kwargs):
+                    self.construct(shaped_args, **kwargs)
+            else:
+                # construct the circuit
+                self.construct(args, **kwargs)
```
For the non-user-facing methods like `construct` and `evaluate`, I would not use the ** operator for kwargs, much like we don't use the * operator for args. Instead we treat args as a tuple, and kwargs as a dict.

josh146 replied:
Fixed

smite commented on pennylane/qnode.py:
```diff
@@ -582,9 +605,15 @@ def jacobian(self, params, which=None, *, method='B', h=1e-7, order=1, **kwargs)
+        circuit_kwargs = {}
+        circuit_kwargs.update(kwargs)
+        for k in ('h', 'order', 'shots', 'force_order2'):
+            circuit_kwargs.pop(k, None)
```
I think there should be a better way of separating the kwargs meant for `QNode.func`, and the ones meant for QNode methods. Maybe `jacobian` should receive the `QNode.func` kwargs as a separate dictionary argument?

josh146 replied:
I don't think this is possible, purely because autograd will always send the function keyword arguments as keyword arguments. What would work is instead to collect the Jacobian keyword arguments within a `grad_options` dictionary.

josh146 replied:
I thought about changing this here, but it seems out of scope. Should we mark this for a new PR?

co9olguy commented:
>     * Since this requires knowledge of the `.val` property, this should be considered an **advanced feature**, available for those who require this code pattern.

Does this actually require knowledge of the `.val` property?

josh146 commented:
> Does this actually require knowledge of the .val property?

Good point. It might not... perhaps I can perform the 'variable unpacking' in the construct method, and pass the direct values to the users quantum function. This would be a one line change --- I'll explore this tomorrow and see if this works/make sure it doesn't break anything.

josh146 commented:
I've added the following:

```python
if self.cache:
    # caching mode, must use variables for kwargs
    # so they can be updated without reconstructing
    res = self.func(*variables, **kwarg_variables)
else:
    # no caching, fine to directly pass kwarg values
    res = self.func(*variables, **keyword_values)
```

This works really nicely, removes the need to use `.val` at all, and leads to a much nicer user experience. I've updated the top comment to reflect this.

co9olguy commented on pennylane/decorator.py:
```diff
@@ -131,12 +131,17 @@ def qnode(device, interface='numpy'):
+
+        cache (bool): If ``True``, the quantum function used to generate the QNode will
+            only be called to construct the quantum circuit once, on first execution,
+            and this circuit will be cached for all further executions. Only activate this
```
```suggestion
            and this circuit structure (i.e., the placement of templates, gates, measurements, etc.) will be cached for all further executions. The circuit parameters can still change with every call. Only activate this
```

co9olguy commented on pennylane/qnode.py:
```diff
@@ -195,16 +195,23 @@ class QNode:
         device (:class:`~pennylane._device.Device`): device to execute the function on
+        cache (bool): If ``True``, the quantum function used to generate the QNode will
+            only be called to construct the quantum circuit once, on first execution,
+            and this circuit will be cached for all further executions. Only activate this
```
update to match any changes from decorator docstring

josh146 replied:
done

co9olguy commented on pennylane/qnode.py:
```diff
@@ -250,20 +257,22 @@ def construct(self, args, **kwargs):
         # flatten the args, replace each with a Variable instance with a unique index
         temp = [Variable(idx) for idx, val in enumerate(_flatten(args))]
         self.num_variables = len(temp)
+        self.args_shape = args
```
this is kind of a weird usage :thinking:

co9olguy replied:
I guess it's meant to be the reference shape, including nested structure?

josh146 replied:
Yep, I just need the shape, but not the actual arguments! There might be a better way of doing this, but we can't assume NumPy here which makes it tricky. Perhaps this is still the cleanest solution?

co9olguy replied:
maybe just use a better name? `args_ref`?

co9olguy replied:
internal flatten function uses the name `model`

co9olguy commented on pennylane/qnode.py:
```diff
@@ -283,14 +292,23 @@ def construct(self, args, **kwargs):
             temp = [Variable(idx, name=key) for idx, _ in enumerate(_flatten(val))]
             kwarg_variables[key] = unflatten(temp, val)
 
+        Variable.free_param_values = np.array(list(_flatten(args)))
```
does `_flatten` not return a list or array??

josh146 replied:
Unfortunately `_flatten` returns a generator (which makes sense for most cases). Since NumPy v1.10 or so, you can no longer instantiate a NumPy array from a generator, hence the intermediate list conversion.

co9olguy commented on pennylane/qnode.py:
```diff
@@ -283,14 +292,23 @@ def construct(self, args, **kwargs):
             temp = [Variable(idx, name=key) for idx, _ in enumerate(_flatten(val))]
             kwarg_variables[key] = unflatten(temp, val)
```
can we just skip this process if `cache=False`?

josh146 replied:
Good point

josh146 replied:
I placed this inside a `if self.cache:` block.

co9olguy commented on pennylane/qnode.py:
```diff
@@ -692,6 +740,12 @@ def _pd_analytic(self, params, idx, force_order2=False, **kwargs):
         Returns:
             float: partial derivative of the node.
         """
+        # remove jacobian specific keyword arguments
```
seems like we might want to put in a small helper function for this, since it is duplicated elsewhere

josh146 replied:
This is just a short term solution though, might be better to just bite the bullet and implement a `grad_options` dictionary. A small change, but it will mean updating a lot of the tests in `test_quantum_gradients` and `test_qnode`.

co9olguy replied:
yep, we might do that. I was thinking of removing duplicated code even now, even if it will end up getting trashed in a future PR anyway

josh146 replied:
done

co9olguy commented on tests/test_qnode.py:
```diff
@@ -559,7 +564,7 @@ def circuit(w):
 
         # circuit jacobians
-        circuit_jacobian = circuit.jacobian(multidim_array)
+        circuit_jacobian = circuit.jacobian([multidim_array])
```
what caused this change?

josh146 replied:
`QNode.jacobian` accepts a list of parameters, so the previous function call was incorrect. Not sure why it passed the tests with caching turned on, but it failed once I turned caching off.

co9olguy commented on tests/test_quantum_gradients.py:
```diff
@@ -396,7 +396,9 @@ def circuit(x,y,z):
-                manualgrad_val = (circuit(angle_inputs + np.pi / 2 * onehot_idx) - circuit(angle_inputs - np.pi / 2 * onehot_idx)) / 2
+                param1 = angle_inputs + np.pi / 2 * onehot_idx
+                param2 = angle_inputs - np.pi / 2 * onehot_idx
+                manualgrad_val = (circuit(*param1) - circuit(*param2)) / 2
```
how come the arguments are now dereferenced?

josh146 replied:
The circuit QNode is defined as `circuit(x, y, z)`, so the arguments should be unpacked before passing. This is similar to the comment above - the existing test _should_ have been failing, not sure why it was passing. The above change allows it to pass with both caching on and off.

co9olguy commented on tests/test_quantum_gradients.py:
```diff
@@ -465,13 +467,13 @@ def d_error(p, grad_method):
             for d_in, d_out in zip(in_data, out_data):
                 args = np.array([d_in, p])
-                diff = (classifier(args) - d_out)
+                diff = (classifier(*args) - d_out)
```
same question

josh146 replied:
The classifier was defined as `classifier_circuit(in_data, x):`, so `args = np.array([d_in, p])` should be unpacked before being passed to the function. Same answer as before - test should not have been passing, don't know why it was previously :thinking:

co9olguy commented on tests/test_quantum_gradients.py:
```diff
@@ -465,13 +467,13 @@ def d_error(p, grad_method):
         y0 = error(param)
         grad = autograd.grad(error)
-        grad_auto = grad([param])
+        grad_auto = grad(param)
```
how come list was removed here?

josh146 replied:
This one confused me for a second, but I remember now: unlike `QNode.jacobian`, `autograd.grad()` accepts unpacked `*args`. As to why it didn't fail in caching mode, but does in caching mode - very good question that I can't answer?

josh146 replied:
^ this change allows it to pass in both caching/no-caching modes, and conforms to the expected `grad` function call.

co9olguy commented:
Also, any new edge cases that we want to check for in new tests?

josh146 commented on pennylane/qnode.py:
```diff
@@ -195,16 +195,23 @@ class QNode:
         device (:class:`~pennylane._device.Device`): device to execute the function on
+        cache (bool): If ``True``, the quantum function used to generate the QNode will
+            only be called to construct the quantum circuit once, on first execution,
+            and this circuit will be cached for all further executions. Only activate this
```
```suggestion
            and this circuit structure (i.e., the placement of templates, gates, measurements, etc.) will be cached for all further executions. The circuit parameters can still change with every call. Only activate this
```

josh146 commented:
I've placed a couple of new tests in `test_qnode.py`, in class `TestQNodeCacheing`. No other edge cases came to mind?
